### PR TITLE
LUT-B quantization accuracy prototype

### PR DESCRIPTION
**Pull request by mgoin**
## Summary

This draft adds reference infrastructure for evaluating NVIDIA Rubin LUT-B
weight compression before a fused kernel is available.

- Defines a canonical logical LUT-B tile as 512 3-bit indices plus an
  eight-entry E4M3 codebook: 200 bytes per 8x64 tile, or 3.125 bits/weight.
- Adds online BF16/FP16 weight quantization for dense linear layers.
- Adds an NVFP4 linear backend that repacks loaded ModelOpt, Quark, and
  compressed-tensors checkpoints in `process_weights_after_loading`.
- Adds selectable calibration-free fitting experiments through
  `quantization_config.linear.algorithm`.
- Uses a deliberately unfused reference forward that reconstructs the weight
  before `torch.nn.functional.linear`.

The persistent representation is intentionally logical and easy to inspect.
An eventual Rubin kernel can swizzle it into the required GMEM/SMEM/TMEM
layout during checkpoint load.

## Motivation

Rubin PTX exposes LUT decompression for the MMA B operand using 3-bit indices
into per-tile E4M3 codebooks. This prototype answers two questions before
kernel work begins:

1. What accuracy is retained by calibration-free BF16-to-LUT-B fitting?
2. Can existing NVFP4 checkpoints be repacked after model load?

Public specification:
https://docs.nvidia.com/cuda/developer-preview/13.4/parallel-thread-execution/index.html#tcgen05-decompress-inp-mat

## Usage

Baseline online BF16-to-LUT-B:

```bash
vllm serve Qwen/Qwen3-8B \
  --quantization lut_b \
  --enforce-eager
```

Calibration-free fitting experiment:

```bash
vllm serve Qwen/Qwen3-8B \
  --quantization online \
  --quantization-config.linear.weight lut_b \
  --quantization-config.linear.algorithm residual_1 \
  --enforce-eager
```

Repack an existing NVFP4 checkpoint:

```bash
vllm serve RedHatAI/Qwen3-8B-NVFP4 \
  --linear-backend lut_b \
  --enforce-eager
```

## Calibration-free GSM8K results

Methodology:

- Model: `Qwen/Qwen3-8B`
- Internal `tests/evals/gsm8k/gsm8k_eval.py` harness
- 1,319 questions, 5-shot
- `max_tokens=256`, `temperature=0`, `seed=42`
- 32 concurrent requests
- B300, eager mode, max model length 8192
- All runs completed 1,319/1,319 HTTP requests successfully

| Algorithm | Row scale | Residuals/tile | Logical bpw | Accuracy |
| --- | ---: | ---: | ---: | ---: |
| baseline | no | 0 | 3.125 | 52.08% |
| `multistart` | no | 0 | 3.125 | 57.62% |
| `scaled` | yes | 0 | 3.131 | 25.78% |
| `residual_1` | no | 1 | 3.174 | 69.83% |
| `scaled_residual_1` | yes | 1 | 3.180 | 66.11% |
| `residual_2` | no | 2 | 3.223 | 72.93% |
| `scaled_residual_2` | yes | 2 | 3.229 | 69.14% |
| `residual_4` | no | 4 | 3.320 | 74.45% |
| `scaled_residual_4` | yes | 4 | 3.327 | 70.13% |
| `residual_8` | no | 8 | 3.516 | 75.51% |
| `scaled_residual_8` | yes | 8 | 3.522 | 74.83% |

Reference checkpoint paths from the same harness:

| Model/path | Activation path | Accuracy |
| --- | --- | ---: |
| BF16 to online LUT-B | BF16 | 52.08% |
| NVFP4 to LUT-B backend | BF16 | 61.03% |
| NVFP4 default | NVFP4 | 86.05% |
| NVFP4 Marlin | BF16 | 87.72% |

The main findings are:

- Deterministic multistart fitting adds 5.54 accuracy points without changing
  the native 3.125-bit representation.
- Weight-space RMS scaling reduced sampled reconstruction error but badly
  regressed model accuracy.
- Preserving one largest-error weight per tile is the best observed
  accuracy/storage knee: 69.83% at 3.174 logical bits/weight.
- Unscaled residual variants win at every matched residual count.

Residual logical accounting uses a 9-bit tile position and BF16 correction.
The reference implementation currently stores the position as `int16`, so its
resident prototype size is slightly larger than the logical checkpoint size.

## Scope and limitations

- This does not execute the Rubin LUT MMA and makes no performance claim.
- Hardware-specific swizzles are intentionally excluded.
- Multistart remains a native 3.125-bit LUT-B representation.
- Per-output scales require an output epilogue.
- Sparse residual sidecars require a separate correction path and are not
  consumed by the LUT MMA.
- MoE weights are not included.

## Validation

```bash
.venv/bin/python -m pytest \
  tests/kernels/quantization/test_nvfp4_lut_b.py \
  tests/quantization/test_online.py \
  tests/quantization/test_quantization_config_args.py -q
```

Result: 33 passed, 7 hardware-gated skips.

Targeted pre-commit hooks, including Ruff, formatting, mypy, markdownlint,
configuration validation, and forbidden-import checks, all pass on the 17
changed files.

## Duplicate-work check

No open `vllm-project/vllm` PR matched searches for `LUT-B`,
`Rubin LUT quantization`, or `tcgen05 decompress`.

## AI assistance disclosure

OpenAI Codex was used to implement, test, evaluate, and prepare this draft.
The human submitter must review every changed line and be able to explain and
defend the implementation and evaluation results before marking it ready for
review.

### CODE REVIEW

**Comment by mgoin** [comment]
Superseded by #51880, which narrows the scope to online quantization of routed MoE expert weights, leaves dense linear layers unquantized, removes the residual/scaling variants, and adds an executable Triton MoE path.
